File: go-backend-interview-prep/09_performance_and_profiling/03_cpu_profiling_pprof_intro/main.py
"""CPU profiling -- Python equivalent of the Go example.

Uses cProfile (stdlib) to profile workloads and pstats to display results.
"""

import cProfile
import hashlib
import pstats
import io


def cpu_intensive_work(iterations: int) -> bytes:
    """Repeated hashing to create measurable CPU load."""
    data = b"hello world performance profiling"
    for _ in range(iterations):
        data = hashlib.sha256(data).digest()
    return data


def string_work(iterations: int) -> str:
    """String manipulation hotspot."""
    s = ""
    for _ in range(iterations):
        s += "x"
        if len(s) > 1000:
            s = s[:100]
    return s


def workload() -> None:
    """Combined workload to profile."""
    cpu_intensive_work(200_000)
    string_work(100_000)


def main() -> None:
    print("=== CPU Profiling with cProfile ===\n")

    # Profile the workload
    profiler = cProfile.Profile()
    profiler.enable()
    workload()
    profiler.disable()

    # Display results
    print("--- Top 10 functions by cumulative time ---\n")
    stream = io.StringIO()
    stats = pstats.Stats(profiler, stream=stream)
    stats.sort_stats("cumulative")
    stats.print_stats(10)
    print(stream.getvalue())

    print("--- Top 10 by total (self) time ---\n")
    stream2 = io.StringIO()
    stats2 = pstats.Stats(profiler, stream=stream2)
    stats2.sort_stats("tottime")
    stats2.print_stats(10)
    print(stream2.getvalue())

    print("--- How to analyze ---")
    print("  python -m cProfile -s cumulative script.py")
    print("  python -m cProfile -o output.prof script.py")
    print()
    print("Key differences from Go pprof:")
    print("  - cProfile is deterministic (counts every call)")
    print("  - pprof is sampling-based (samples at intervals)")
    print("  - cProfile has higher overhead -- not for production")
    print("  - Go's pprof is production-safe (sampling = low overhead)")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/09_performance_and_profiling/03_cpu_profiling_pprof_intro/main.go
package main

import (
	"crypto/sha256"
	"fmt"
	"os"
	"runtime/pprof"
	"time"
)

// cpuIntensiveWork does repeated hashing to create a measurable CPU load.
func cpuIntensiveWork(iterations int) [32]byte {
	data := []byte("hello world performance profiling")
	var hash [32]byte
	for i := 0; i < iterations; i++ {
		hash = sha256.Sum256(data)
		data = hash[:]
	}
	return hash
}

// stringWork does string manipulation to show up as a separate hotspot.
func stringWork(iterations int) string {
	s := ""
	for i := 0; i < iterations; i++ {
		s += "x"
		if len(s) > 1000 {
			s = s[:100]
		}
	}
	return s
}

func main() {
	// --- Create CPU profile ---
	f, err := os.Create("cpu.prof")
	if err != nil {
		fmt.Println("could not create profile:", err)
		return
	}
	defer f.Close()

	fmt.Println("=== CPU Profiling with pprof ===")
	fmt.Println("Starting CPU profile -> cpu.prof")

	if err := pprof.StartCPUProfile(f); err != nil {
		fmt.Println("could not start profile:", err)
		return
	}
	defer pprof.StopCPUProfile()

	// --- Run workloads ---
	start := time.Now()

	fmt.Println("\nRunning CPU-intensive hashing...")
	hash := cpuIntensiveWork(500_000)
	fmt.Printf("  hash result: %x... (%s)\n", hash[:4], time.Since(start))

	mid := time.Now()
	fmt.Println("Running string manipulation...")
	s := stringWork(100_000)
	fmt.Printf("  string len: %d (%s)\n", len(s), time.Since(mid))

	fmt.Printf("\nTotal: %s\n", time.Since(start))

	// Profile is written when StopCPUProfile runs (deferred)
	fmt.Println("\n--- How to analyze ---")
	fmt.Println("  go tool pprof cpu.prof")
	fmt.Println()
	fmt.Println("Key pprof commands:")
	fmt.Println("  top10         -- show top 10 functions by flat time")
	fmt.Println("  top -cum      -- sort by cumulative time")
	fmt.Println("  list funcName -- show annotated source for a function")
	fmt.Println("  web           -- open interactive SVG graph in browser")
	fmt.Println("  peek funcName -- show callers and callees")
	fmt.Println("  quit          -- exit pprof")
	fmt.Println()
	fmt.Println("For live services, import net/http/pprof:")
	fmt.Println("  import _ \"net/http/pprof\"")
	fmt.Println("  // Then: go tool pprof http://localhost:6060/debug/pprof/profile?seconds=10")
}
